plugins/in_docker_events/docker_events.c

```diff
@@ -153,6 +155,39 @@ static int in_de_collect(struct flb_input_instance *ins,
             }
         }
     }
+    else if (ret == 0) {
+        /* EOF */
+        /* docker service may be restarted */
+        flb_plg_info(ctx->ins, "EOF detected. Re-initialize");
+        /* remove old socket collector */
+        ret = flb_input_collector_delete(ctx->coll_id, ins);
+        close(ctx->fd);
+        if (ret < 0) {
+            flb_plg_error(ctx->ins, "failed to pause event");
+            return -1;
+        }
+        ctx->fd = -1;
+        /* create socket again */
```

Comment:
Hi @nokute78 ,

Thanks so much for proposing this fix. 
I have tried to play with your PR in my machine, however I have seen this error after rebooting docker daemon. 
Seems the first retry failed and it never re-try again.
I have encountered similar issues when trying different workarounds on my side, not sure how to resolve this.

```
~/nokute78/fluent-bit/build$ sudo bin/fluent-bit -i docker_events -o stdout
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/05/03 17:17:15] [ info] [engine] started (pid=3592)
[2021/05/03 17:17:15] [ info] [storage] version=1.1.1, initializing...
[2021/05/03 17:17:15] [ info] [storage] in-memory
[2021/05/03 17:17:15] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/05/03 17:17:15] [ info] [input:docker_events:docker_events.0] listening for events on /var/run/docker.sock
[2021/05/03 17:17:15] [ info] [sp] stream processor started
[2021/05/03 17:17:24] [ info] [input:docker_events:docker_events.0] EOF detected. Re-initialize
[2021/05/03 17:17:24] [error] [plugins/in_docker_events/docker_events.c:71 errno=104] Connection reset by peer
[2021/05/03 17:17:24] [ info] [input:docker_events:docker_events.0] EOF detected. Re-initialize
[2021/05/03 17:17:24] [error] [plugins/in_docker_events/docker_events.c:59 errno=111] Connection refused
[2021/05/03 17:17:24] [error] [input:docker_events:docker_events.0] failed to re-initialize socket
```

Reply:
Could you check docker/systemd logs ?

> [plugins/in_docker_events/docker_events.c:71 errno=104] Connection reset by peer

This means destination resets the connection. (e.g. docker or linux kernel)

FYI, Apparmor causes this errno in this case.
https://github.com/moby/moby/issues/14504

Reply:
Thanks @nokute78  for the hint, I didn't see any errors in the dockerd logs though.

~/nokute78/fluent-bit/build$ sudo bin/fluent-bit -i docker_events -o stdout
```
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/05/03 23:24:44] [ info] [engine] started (pid=10336)
[2021/05/03 23:24:44] [ info] [storage] version=1.1.1, initializing...
[2021/05/03 23:24:44] [ info] [storage] in-memory
[2021/05/03 23:24:44] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/05/03 23:24:44] [ info] [input:docker_events:docker_events.0] listening for events on /var/run/docker.sock
[2021/05/03 23:24:44] [ info] [sp] stream processor started
[2021/05/03 23:25:08] [ info] [input:docker_events:docker_events.0] EOF detected. Re-initialize
[2021/05/03 23:25:08] [error] [plugins/in_docker_events/docker_events.c:71 errno=104] Connection reset by peer
[2021/05/03 23:25:08] [ info] [input:docker_events:docker_events.0] EOF detected. Re-initialize
[2021/05/03 23:25:08] [error] [plugins/in_docker_events/docker_events.c:59 errno=111] Connection refused
[2021/05/03 23:25:08] [error] [input:docker_events:docker_events.0] failed to re-initialize socket
```
~/nokute78/fluent-bit/build$ systemctl status docker
```
● docker.service - Docker Application Container Engine
   Loaded: loaded (/lib/systemd/system/docker.service; enabled; vendor preset: enabled)
  Drop-In: /etc/systemd/system/docker.service.d
           └─override.conf
   Active: active (running) since Mon 2021-05-03 23:25:10 UTC; 7s ago
     Docs: https://docs.docker.com
  Process: 10591 ExecStartPre=/bin/sh -c  cd /datadrive   (code=exited, status=0/SUCCESS)
 Main PID: 10594 (dockerd)
    Tasks: 26
   CGroup: /system.slice/docker.service
           └─10594 /usr/bin/dockerd -H fd:// --containerd /var/run/containerd/containerd.sock

May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334860190Z" level=warning msg="Your kernel does not support cgroup blkio weight"
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334867790Z" level=warning msg="Your kernel does not support cgroup blkio weight_device"
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.335061591Z" level=info msg="Loading containers: start."
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: fluent#New: AsyncConnect is now deprecated, please use Async insteadfluent#New: AsyncConnect is now deprecated, please use Async insteadfluent#New: AsyncConnect is now deprecated, please use Async insteadtime="2021-05-03T23:25:10.163716322Z" level=info msg="Ther
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.207732489Z" level=info msg="Loading containers: done."
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.411914628Z" level=warning msg="Not using native diff for overlay2, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled" storage-driver=overlay2
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.412652133Z" level=info msg="Docker daemon" commit=8728dd246c3ab53105434eef8ffe997b6fd14dc6 graphdriver(s)=overlay2 version=20.10.6+azure
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.412991435Z" level=info msg="Daemon has completed initialization"
May 03 23:25:10 c22f0b9ef342 systemd[1]: Started Docker Application Container Engine.
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.457327504Z" level=info msg="API listen on /var/run/docker.sock"
```

~/nokute78/fluent-bit/build$ journalctl -u docker.service
```
journalctl -u docker.service
-- Logs begin at Mon 2021-05-03 23:21:54 UTC, end at Mon 2021-05-03 23:25:35 UTC. --
May 03 23:25:08 c22f0b9ef342 systemd[1]: Stopping Docker Application Container Engine...
May 03 23:25:08 c22f0b9ef342 dockerd[5305]: time="2021-05-03T23:25:08.899503647Z" level=info msg="Processing signal 'terminated'"
May 03 23:25:08 c22f0b9ef342 dockerd[5305]: time="2021-05-03T23:25:08.905368783Z" level=info msg="Daemon shutdown complete"
May 03 23:25:08 c22f0b9ef342 dockerd[5305]: time="2021-05-03T23:25:08.905514284Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=plugins.moby
May 03 23:25:08 c22f0b9ef342 systemd[1]: Stopped Docker Application Container Engine.
May 03 23:25:08 c22f0b9ef342 systemd[1]: Starting Docker Application Container Engine...
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.021033485Z" level=info msg="Starting up"
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.023844002Z" level=info msg="detected 127.0.0.53 nameserver, assuming systemd-resolved, so using resolv.conf: /run/systemd/resolve/resolv.conf"
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.032957858Z" level=info msg="parsed scheme: \"unix\"" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.032998058Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.033043058Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.033062258Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034334066Z" level=info msg="parsed scheme: \"unix\"" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034361266Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034382166Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034402466Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334826690Z" level=warning msg="Your kernel does not support CPU realtime scheduler"
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334860190Z" level=warning msg="Your kernel does not support cgroup blkio weight"
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334867790Z" level=warning msg="Your kernel does not support cgroup blkio weight_device"
May 03 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.335061591Z" level=info msg="Loading containers: start."
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: fluent#New: AsyncConnect is now deprecated, please use Async insteadfluent#New: AsyncConnect is now deprecated, please use Async insteadfluent#New: AsyncConnect is now deprecated, please use Async insteadtime="2021-05-03T23:25:10.163716322Z" level=info msg="Ther
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.207732489Z" level=info msg="Loading containers: done."
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.411914628Z" level=warning msg="Not using native diff for overlay2, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled" storage-driver=overlay2
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.412652133Z" level=info msg="Docker daemon" commit=8728dd246c3ab53105434eef8ffe997b6fd14dc6 graphdriver(s)=overlay2 version=20.10.6+azure
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.412991435Z" level=info msg="Daemon has completed initialization"
May 03 23:25:10 c22f0b9ef342 systemd[1]: Started Docker Application Container Engine.
May 03 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.457327504Z" level=info msg="API listen on /var/run/docker.sock"
```

Reply:
Thank you for quick response. 
Hmm...

1.  can you share `dmesg` ?
2. Is there any logs before restarting at `/var/log` ?

`systemctl` and `journalctl` logs seems to be logs after restarting docker.
In my environment, `/var/log/syslog` is it.

```
May  4 08:42:27 locals dockerd[14165]: time="2021-05-04T08:42:27.266660436+09:00" level=info msg="Processing signal 'terminated'"
May  4 08:42:27 locals dockerd[14165]: time="2021-05-04T08:42:27.268357743+09:00" level=info msg="Daemon shutdown complete"
May  4 08:42:27 locals systemd[1]: Stopping Docker Application Container Engine...
May  4 08:42:27 locals systemd[1]: docker.service: Succeeded.
May  4 08:42:27 locals systemd[1]: Stopped Docker Application Container Engine.
May  4 08:42:27 locals systemd[1]: Starting Docker Application Container Engine...
May  4 08:42:27 locals dockerd[14804]: time="2021-05-04T08:42:27.342604910+09:00" level=info msg="Starting up"
May  4 08:42:27 locals dockerd[14804]: time="2021-05-04T08:42:27.344636256+09:00" level=info msg="detected 127.0.0.53 nameserver, assuming systemd-resolved, so using resolv.conf: /run/systemd/resolve/resolv.conf"
```

Reply:
This is the dmesg output, will check in syslog.
[dmesg.txt](https://github.com/fluent/fluent-bit/files/6418409/dmesg.txt)

Reply:
This is the syslog:
```
May  3 23:25:08 c22f0b9ef342 systemd[1]: Stopping Docker Application Container Engine...
May  3 23:25:08 c22f0b9ef342 dockerd[5305]: time="2021-05-03T23:25:08.899503647Z" level=info msg="Processing signal 'terminated'"
May  3 23:25:08 c22f0b9ef342 dockerd[5305]: time="2021-05-03T23:25:08.905368783Z" level=info msg="Daemon shutdown complete"
May  3 23:25:08 c22f0b9ef342 dockerd[5305]: time="2021-05-03T23:25:08.905514284Z" level=info msg="stopping event stream following graceful shutdown" error="context canceled" module=libcontainerd namespace=plugins.moby
May  3 23:25:08 c22f0b9ef342 systemd[1]: Stopped Docker Application Container Engine.
May  3 23:25:08 c22f0b9ef342 systemd[1]: Closed Docker Socket for the API.
May  3 23:25:08 c22f0b9ef342 systemd[1]: Stopping Docker Socket for the API.
May  3 23:25:08 c22f0b9ef342 systemd[1]: Starting Docker Socket for the API.
May  3 23:25:08 c22f0b9ef342 systemd[1]: Listening on Docker Socket for the API.
May  3 23:25:08 c22f0b9ef342 systemd[1]: Starting Docker Application Container Engine...
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.021033485Z" level=info msg="Starting up"
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.023844002Z" level=info msg="detected 127.0.0.53 nameserver, assuming systemd-resolved, so using resolv.conf: /run/systemd/resolve/resolv.conf"
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.032957858Z" level=info msg="parsed scheme: \"unix\"" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.032998058Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.033043058Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.033062258Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034334066Z" level=info msg="parsed scheme: \"unix\"" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034361266Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034382166Z" level=info msg="ccResolverWrapper: sending update to cc: {[{unix:///var/run/containerd/containerd.sock  <nil> 0 <nil>}] <nil> <nil>}" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.034402466Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334826690Z" level=warning msg="Your kernel does not support CPU realtime scheduler"
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334860190Z" level=warning msg="Your kernel does not support cgroup blkio weight"
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.334867790Z" level=warning msg="Your kernel does not support cgroup blkio weight_device"
May  3 23:25:09 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:09.335061591Z" level=info msg="Loading containers: start."
May  3 23:25:10 c22f0b9ef342 dockerd[10594]: fluent#New: AsyncConnect is now deprecated, please use Async insteadfluent#New: AsyncConnect is now deprecated, please use Async insteadfluent#New: AsyncConnect is now deprecated, please use Async insteadtime="2021-05-03T23:25:10.163716322Z" level=info msg="There are old running containers, the network config will not take affect"
May  3 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.207732489Z" level=info msg="Loading containers: done."
May  3 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.411914628Z" level=warning msg="Not using native diff for overlay2, this may cause degraded performance for building images: kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled" storage-driver=overlay2
May  3 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.412652133Z" level=info msg="Docker daemon" commit=8728dd246c3ab53105434eef8ffe997b6fd14dc6 graphdriver(s)=overlay2 version=20.10.6+azure
May  3 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.412991435Z" level=info msg="Daemon has completed initialization"
May  3 23:25:10 c22f0b9ef342 systemd[1]: Started Docker Application Container Engine.
May  3 23:25:10 c22f0b9ef342 dockerd[10594]: time="2021-05-03T23:25:10.457327504Z" level=info msg="API listen on /var/run/docker.sock"
```

Reply:
I sent a patch to fix this `Connection reset by peer` issue.
Please test `reconnect.retry_limits` and `reconnect.retry_interval`.